proposed/MPID Messaging.md

```diff
@@ -0,0 +1,295 @@
+    type_tag: MPID_MESSAGE_TAG,
+    identifier: mpid_message_name(mpid_message),
+    data: encode(mpid_message),
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![mpid_message.sender],
+    previous_owner_signatures: vec![]
+}
+```
+```rust
+let sd_for_mpid_header = StructuredData {
+    type_tag: MPID_HEADER_TAG,
+    identifier: mpid_message_name(mpid_message),  // or mpid_header_name(signed_header)
+    data: mpid_message.signed_header,  // or inbox.signed_header
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![mpid_message.sender],  // or vec![inbox.sender]
+    previous_owner_signatures: vec![]
+}
+```
+
+## Message Flow
+
+```
+                    MpidManagers(A)    MpidManagers(B)
+                   /      *                   *        \
+MpidClient(A) ->  -       *                   *         -  <- MpidClient(B)
+                   \      *                   *        /
+
+```
+1. `MpidClient(A)` sends an `MpidMessage` to `MpidManagers(A)` as a Put request.
+1. `MpidManagers(A)` store this message in their outbox for A and send the `signed_header` component to `MpidManagers(B)` as a Put request, unless the outbox is full in which case a PutFailure response is returned to `MpidClient(A)`.
+1. `MpidManagers(B)` try to store the `(sender, signed_header)` in their inbox for B.  If successful, they forward it to `MpidClient(B)` as a Put request immediately or as soon as it appears online.  If unsuccessful (e.g. the inbox is full or sender has been blacklisted), they reply to `MpidManagers(A)` with a PutFailure, who then remove the message from the outbox and send a PutFailure response to `MpidClient(A)`.
+1. To retrieve the message from the network, `MpidClient(B)` sends a Get request to `MpidManagers(A)`.
+1. `MpidManagers(A)` send a GetResponse to `MPidClient(B)`.
+1. On receiving the message, `MpidClient(B)` sends a remove request to `MpidManagers(B)` via a Delete.
+1. `MpidManagers(B)` remove the corresponding entry from the inbox for B and forward the remove request to `MpidManagers(A)`.  `MpidManagers(A)` then remove the corresponding entry from the outbox for A.
+
+`MpidClient(A)` might decide to remove the `MpidMessage` from the outbox if the message hasn't been retrieved by `MpidClient(B)` yet.  In this case, `MpidManagers(A)` should not only remove the corresponding `MpidMessage` from their outbox for A, but also send a notification to the group of `MpidManagers(B)` so they can remove the corresponding entry from their inbox of B.  These messages will all be Delete requests.
+
+_MPidClient(A)_ =>> |__MPidManagers(A)__ (Put)(Header.So) *->> | __MPidManagers(B)__  (Store(Header))(Online(MpidClient(B)) ? Header.So : (WaitForOnlineB)(header.So)) *-> | _MpidClient(B)_ So.Retrieve ->> | __MpidManagers(A)__ So.Message *-> | _MpidClient(B)_ Remove.So ->> | __MpidManagers(B)__ {Remove(Header), Remove.So} *->> | __MpidManagers(A)__ Remove
```

Comment:
I've been meaning to get round to adding a flowchart for the message flow, since this may be easier to read than either of the two formats above?  What do you think @ustulation ?